go-controller/pkg/libovsdbops/db_object_ids.go

```diff
@@ -0,0 +1,283 @@
+	objectIDs map[ExternalIDKey]string
+}
+
+// NewDbObjectIDs is used to construct DbObjectIDs, idsType and controller are always required,
+// objectIds may be empty, or half-filled for predicate search.
+// objectIds keys that are not used by given idsType will not be used.
+func NewDbObjectIDs(idsType ObjectIDsType, controller string, objectIds map[ExternalIDKey]string) *DbObjectIDs {
+	if controller == "" {
+		panic("NewDbObjectIDs failed: controller should not be empty")
+	}
+	externalIDKeys := idsType.GetExternalIDKeys()
+	if externalIDKeys == nil {
+		// can only happen if ObjectIDsType{} is passed
+		panic(fmt.Sprintf("NewDbObjectIDs failed: ObjectIDsType %v should not be empty", idsType))
+	}
+	// only use values for keys from idsType
+	filteredObjectIDs := map[ExternalIDKey]string{}
+	for _, externalIDKey := range externalIDKeys {
+		if value, ok := objectIds[externalIDKey]; ok {
+			filteredObjectIDs[externalIDKey] = value
+		}
+	}
+	objectIDs := &DbObjectIDs{
+		idsType:             idsType,
+		ownerControllerName: controller,
+		objectIDs:           filteredObjectIDs,
+	}
+	return objectIDs
+}
+
+func (objectIDs *DbObjectIDs) NewDbObjectWithExtraIDs(extraObjectIds map[ExternalIDKey]string) *DbObjectIDs {
+	ids := deepcopyMap(objectIDs.objectIDs)
+	for key, value := range extraObjectIds {
+		ids[key] = value
+	}
+	return NewDbObjectIDs(objectIDs.idsType, objectIDs.ownerControllerName, ids)
+}
+
+func (objectIDs *DbObjectIDs) NewDbObjectWithoutIDs(idsToDelete ...ExternalIDKey) *DbObjectIDs {
+	ids := deepcopyMap(objectIDs.objectIDs)
+	for _, keyToDel := range idsToDelete {
+		delete(ids, keyToDel)
+	}
+	return NewDbObjectIDs(objectIDs.idsType, objectIDs.ownerControllerName, ids)
+}
+
+func (objectIDs *DbObjectIDs) HasSameOwner(ownerController string, indexT ObjectIDsType) bool {
+	return objectIDs.ownerControllerName == ownerController && objectIDs.idsType.IsSameType(indexT)
+}
+
+// GetObjectID returns value from objectIDs.objectIDs map, and empty string for not found values.
+// Usually objectIDs.objectIDs doesn't include PrimaryIDKey, OwnerTypeKey, and OwnerControllerKey.
+func (objectIDs *DbObjectIDs) GetObjectID(key ExternalIDKey) string {
+	return objectIDs.objectIDs[key]
+}
+
+// GetExternalIDs returns a map of ids, that always includes keys
+// - OwnerControllerKey
+// - OwnerTypeKey
+// - PrimaryIDKey (constructed by DbObjectIDs.GetUniqueID)
+// and also all keys that are used by objectIDs.idsType.
+// If some required keys are not present in the DbObjectIDs.objectIDs, they will be replaced with empty strings.
+// That may lead to having entries with the same primary id, so make sure to always set all required ids before
+// calling this method unless empty ids are intended.
+func (objectIDs *DbObjectIDs) GetExternalIDs() map[string]string {
+	externalIDs := objectIDs.getExternalIDsWithoutPrimary(false)
+	externalIDs[PrimaryIDKey.String()] = objectIDs.GetUniqueID()
+	return externalIDs
```

Comment:
And perhaps we should not set PrimaryIDKey if GetUniqueID return an empty id or error.

Reply:
the common note for all the other comments around error handling (since it is all related)
we decided to simplify verification and thus error handling, by only checking `idsType` and `ownerControllerName`, and letting callers set `objectIDs` as they wish (there is even a strange use case for it now, where the `ovnAddressSetFactory` uses `DbObjectIDs.GetUniqueID` when `ipFamily` is not set to create unique key without ipFamily)
That is also the reason why we don't return an error in `getUniqueIDWithCheck` immediately with empty string, but construct id based on the fields that are set. Since it is easy to figure out which ids should be set for every `ObjectIDsType` I hope that shouldn't be a problem.
But we can re-consider this approach is we want more verification

Reply:
One thing is not producing an error so you don't have any error handling. But another thing is setting/providing an incorrect `id`. 
They are different things. We can discuss it but, if the user were behaving responsible, what would break if we did not set an id when we don't have the required composing ids?

Reply:
I guess in that case we should only set ids that are present in `objectIds`, and not replace them with empty strings (that is probably what you meant initially, but I was confused by the emphasis on the primary id)